File: week 3/different_summands/different_summands.cpp
#include <iostream>
#include <vector>
#include <cassert>
using namespace std;

vector<int> optimal_summands(int n) {
  vector<int> summands;

  if (n <= 2) {
    summands.push_back(n);
  } else {
    int l = 1;

    while (n > 2 * l) {
      summands.push_back(l);
      n -= l++;
    }

    if (n > 0) {
      summands.push_back(n);
    }
  }

  return summands;
}

void test_solution() {
  vector<int> result;

  result = optimal_summands(2);
  assert(result.size() == 1);
  assert(result[0] == 2);
  result = optimal_summands(5);
  assert(result.size() == 2);
  assert(result[0] == 1);
  assert(result[1] == 4);
  result = optimal_summands(8);
  assert(result.size() == 3);
  assert(result[0] == 1);
  assert(result[1] == 2);
  assert(result[2] == 5);
}

int main() {
  test_solution();

  int n;
  std::cin >> n;
  vector<int> summands = optimal_summands(n);
  std::cout << summands.size() << '\n';
  for (size_t i = 0; i < summands.size(); ++i) {
    std::cout << summands[i] << ' ';
  }
}

File: week 3/different_summands/different_summands.py
# Uses python3
import sys

def optimal_summands(n):
    if n <= 2:
        return [n]

    l = 1
    summands = []

    while n > 2 * l:
        summands.append(l)
        n -= l
        l += 1

    if n > 0:
        summands.append(n)

    return summands

def test_solution():
    are_lists_identical(optimal_summands(2), [2])
    are_lists_identical(optimal_summands(5), [1, 4])
    are_lists_identical(optimal_summands(6), [1, 2, 3])
    are_lists_identical(optimal_summands(8), [1, 2, 5])

def are_lists_identical(l1, l2):
    assert len(set(l1).union(l2)) == len(l1)

if __name__ == '__main__':
    test_solution()

    input = sys.stdin.read()
    n = int(input)
    summands = optimal_summands(n)
    print(len(summands))
    for x in summands:
        print(x, end=' ')
